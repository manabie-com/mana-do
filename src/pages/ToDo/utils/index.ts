import { Todo, TodoStatus } from '../../../models/todo';
const created_date = new Date().toISOString();
export const defaultToDos = {
  content: 'Content',
  created_date,
  status: TodoStatus.ACTIVE,
  id: 'testId',
  user_id: 'firstUser',
} as Todo;

export function spawnToDos(count: number): Todo[] {
  return new Array(count).fill(defaultToDos).map(todo => {
    todo.id = `testId-${Math.random()}`;
    return todo;
  });
}
