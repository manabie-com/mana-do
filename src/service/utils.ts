import {Todo, TodoStatus} from "../models/todo";

export interface TResponse {
    status: Number,
    code: string,
    data?: any
}

export const updateTodoService = async (todoId:string, content: string): Promise<TResponse> => {

    const dataListTodo:any = await localStorage.getItem('todolist') || '[]';

    const dataToAPI = await JSON.parse(dataListTodo) as Todo[]

    const index = dataToAPI.findIndex(todo => todo.id === todoId)

    dataToAPI[index].content = content

    await localStorage.setItem('todolist', JSON.stringify(dataToAPI))

    return {status: 200, code: "success"}

}

export const removeTodoService = async (todoId: string): Promise<TResponse> => {
    const dataListTodo:any = await localStorage.getItem('todolist') || '[]';

    const dataToAPI = await JSON.parse(dataListTodo) as Todo[]

    const index = dataToAPI.findIndex(todo => todo.id === todoId)

    dataToAPI.splice(index, 1)

    await localStorage.setItem('todolist', JSON.stringify(dataToAPI))

    return {status: 200, code: "success"}

}

export const updateStatusTodoService = async (todoId:string, status: TodoStatus): Promise<TResponse> => {

    const dataListTodo:any = await localStorage.getItem('todolist') || '[]';

    const dataToAPI = await JSON.parse(dataListTodo) as Todo[]

    const index = dataToAPI.findIndex(todo => todo.id === todoId)

    dataToAPI[index].status = status

    await localStorage.setItem('todolist', JSON.stringify(dataToAPI))

    return {status: 200, code: "success"}

}