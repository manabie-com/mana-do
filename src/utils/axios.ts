import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
const axiosClient = axios.create({
  baseURL: 'https://61615cd137492500176312cc.mockapi.io/',
  headers: {
    'Content-Type': 'application/json',
  },
});
// Add a request interceptor
axiosClient.interceptors.request.use(
  function (config: AxiosRequestConfig) {
    // Do something before request is sent
    return config;
  },
  function (error) {
    // Do something with request error
    return Promise.reject(error);
  }
);

// Add a response interceptor
axiosClient.interceptors.response.use(
  function (response: AxiosResponse) {
    // Any status code that lie within the range of 2xx cause this function to trigger
    // Do something with response data
    return response;
  },
  function (error) {
    // Any status codes that falls outside the range of 2xx cause this function to trigger
    // Do something with response error
    return Promise.reject(error);
  }
);
export default axiosClient;