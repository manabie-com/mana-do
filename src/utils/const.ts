export const TO_DO_LIST = "TO_DO_LIST";