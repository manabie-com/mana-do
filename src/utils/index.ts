import { Todo, TodoStatus } from '../models/todo';

export function isTodoCompleted(todo: Todo): boolean {
  return todo.status === TodoStatus.COMPLETED;
}

export function isTodoActive(todo: Todo): boolean {
  return todo.status === TodoStatus.ACTIVE;
}
