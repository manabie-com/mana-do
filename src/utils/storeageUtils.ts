import {initialState} from '../store/reducer';
import {Todo, TodoStatus} from '../models/todo';
import {createTodo, deleteTodo, updateTodoContent, updateTodoStatus} from "./todoUtils";

export const storeLoginToken = (token: string) => localStorage.setItem('token', token);

export const removeLoginToken = () => localStorage.removeItem('token');

export const getTokenStorage = () => localStorage.getItem('token');

export const storeReducerStorage = (data: object) => localStorage.setItem('reducerLocalStorage', JSON.stringify(data));

export const getReducerFromStorage = () => localStorage.getItem('reducerLocalStorage');

export const getTodosFromStorage = () => {
  const reducer = getReducerFromStorage();
  if (reducer) {
    const { todos } = JSON.parse(reducer);

    return todos;
  }

  return [];
};

export const storeTodos = (todoItem: Todo) => {
  const reducer = getReducerFromStorage();

  if (!reducer) {
    const newReducer = {
      todos: [todoItem]
    }
    storeReducerStorage(newReducer);
  } else {
    const parsed = JSON.parse(reducer);
    const newData = createTodo(parsed, todoItem);
    storeReducerStorage(newData);
  }
}

export const clearTodoStorage = () => {
  const reducer = getReducerFromStorage();
  if (reducer) {
    const parsed = JSON.parse(reducer);
    const newReducer = {
      ...parsed,
      todos: []
    }
    storeReducerStorage(newReducer)
  }
}

export const storeTodosStatus = (todoId: string, isChecked: boolean) => {
  const reducer = getReducerFromStorage();
  if (reducer) {
    const parsed = JSON.parse(reducer);
    storeReducerStorage(updateTodoStatus(parsed, todoId, isChecked))
  }
}

export const storeToggleAllTodosStatus = (checked: boolean) => {
  const reducer = getReducerFromStorage();
  if (reducer) {
    const parsed = JSON.parse(reducer);
    const tempTodos = parsed.todos.map((todo: Todo) => ({
      ...todo,
      status: checked ? TodoStatus.Completed : TodoStatus.Active
    }));
    const data = {
      ...parsed,
      todos: tempTodos
    }
    storeReducerStorage(data)
  }
}

export const storeTodosContent = (todoId: string, content: string) => {
  const reducer = getReducerFromStorage();
  if (reducer) {
    const parsed = JSON.parse(reducer);
    const newTodo = updateTodoContent(parsed.todos, todoId, content);
    const data = {
      ...parsed,
      todos: newTodo
    }
    storeReducerStorage(data);
  }
}

export const storeDeleteTodos = (todoId: string) => {
  const reducer = getReducerFromStorage();
  if (reducer) {
    const parsed = JSON.parse(reducer);
    storeReducerStorage(deleteTodo(parsed, todoId))
  }
}

export const initializerState = (initialValue = initialState) => {
  const localReducerData = getReducerFromStorage();
  if (localReducerData) {
    return JSON.parse(localReducerData);
  }
  storeReducerStorage(initialValue);
  return initialValue;
}
