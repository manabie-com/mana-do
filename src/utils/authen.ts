import {getTokenStorage} from './storeageUtils';

export const isAuthenticated = () => Boolean(getTokenStorage());
