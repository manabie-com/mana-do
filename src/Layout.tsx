import React from 'react';
import {isAuthenticated} from "./utils/authen";
import {removeLoginToken} from "./utils/storeageUtils";
import {useHistory} from "react-router-dom";
import {ROUTES} from "./utils/constants";

interface LayoutProps {
  children: React.ReactNode;
}

const Layout = ({children}: LayoutProps) => {
  const history = useHistory();

  const onLogOut =() => {
    removeLoginToken();
    history.replace(ROUTES.SIGN_IN);
  }

  return (
    <>
      <nav className="main-nav">
        <div className="nav-inner">
          {
            isAuthenticated() ? <button onClick={onLogOut} className="basic red">Logout</button> : (
              <>
                <button className="basic green">Login</button>
                <button>Sign up</button>
              </>
            )
          }
        </div>
      </nav>
      <section>{children}</section>
    </>
  )
}

export default Layout;
