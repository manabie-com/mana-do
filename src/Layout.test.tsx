import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Layout from "./Layout";
import {getTokenStorage, removeLoginToken, storeLoginToken} from "./utils/storeageUtils";

const MockLayoutChildren = () => {
  return (
    <div>Layout Children</div>
  )
};

describe('<Layout />', () => {
  test('should return snapshot logged-in', () => {
    storeLoginToken('token');
    const tree = renderer.create(<Layout><MockLayoutChildren/></Layout>).toJSON();
    expect(tree).toMatchSnapshot();
  })

  test('should return snapshot not logged-in yet', () => {
    removeLoginToken();
    const tree = renderer.create(<Layout><MockLayoutChildren/></Layout>).toJSON();
    expect(tree).toMatchSnapshot();
  })
})

describe('testing functions', () => {
  test('should remove token after logout success', () => {
    storeLoginToken('token');
    const wrapper = shallow(
      <Layout>
        <MockLayoutChildren/>
      </Layout>
    );
    const logoutButton = wrapper.find('button');
    const logoutButtonProps = logoutButton.props();

    // on logout clicked
    // @ts-ignore
    logoutButtonProps.onClick();
    expect(getTokenStorage()).toBeNull();
  })
})
