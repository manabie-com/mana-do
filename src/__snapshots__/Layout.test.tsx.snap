// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> should return snapshot logged-in 1`] = `
Array [
  <nav
    className="main-nav"
  >
    <div
      className="nav-inner"
    >
      <button
        className="basic red"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </nav>,
  <section>
    <div>
      Layout Children
    </div>
  </section>,
]
`;

exports[`<Layout /> should return snapshot not logged-in yet 1`] = `
Array [
  <nav
    className="main-nav"
  >
    <div
      className="nav-inner"
    >
      <button
        className="basic green"
      >
        Login
      </button>
      <button>
        Sign up
      </button>
    </div>
  </nav>,
  <section>
    <div>
      Layout Children
    </div>
  </section>,
]
`;
